Refactor geopmdpy/session.py and add doc strings

cmcantalupo opened the pull request:
- Fixes #1711 from github issues.

bgeltz commented on service/geopmdpy/session.py:
```diff
@@ -32,90 +32,237 @@
+            domain_idx = int(words[2])
+            self._requests.append((signal_name, domain_type, domain_idx))
+
+    def init_signal_format(self):
```
Some docstrings would help explain what's going on here.

Why is the 0th request (of potentially multiple?) being paried with the 4th value back from SignalInfo?  I assume this is to extract the format function?

What if there are multiple requests?

cmcantalupo replied:
each request is a tuple (signal_name, domain_type, domain_idx).  Later we use the signal_name (the 0th element of the request) to get back an info tuple.  Element 4 of the info tuple is the format enum.

bgeltz reviewed (changes requested):
Mostly LGTM.  1 question.

cmcantalupo committed:
Add C interface that will format a signal based on the format enum

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add python wrapper for geopm_pio_format_signal

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Complete refactor of geopmsession implementation

- Account for read overhead in timed loop.
- Apply prescribed formatting for printed signals.
- Split implemenation into three testable objects:
  Session, TimedLoop and Requests

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
A couple of fixups for access.py and service.py

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Clean up TimedLoop example

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz commented on service/src/PlatformIO.cpp:
```diff
@@ -967,4 +967,28 @@ extern "C" {
+    int geopm_pio_format_signal(double signal,
+                                int format_type,
+                                size_t result_max,
+                                char *result)
```
result needs to be checked for NULL.

https://en.cppreference.com/w/c/string/byte/strncpy
The behavior is undefined ... if either dest or src is not a pointer to a character array (including if dest or src is a null pointer)

cmcantalupo replied:
https://github.com/geopm/geopm/issues/1778

bgeltz commented on service/src/PlatformIO.cpp:
```diff
@@ -967,4 +967,28 @@ extern "C" {
+            auto format_func = geopm::string_format_type_to_function(format_type);
+            std::string result_cxx = format_func(signal);
+            result[result_max - 1] = '\0';
+            strncpy(result, result_cxx.c_str(), result_max);
```
We need to open an issue about changing all strncpy's to strncpy_s's.

bgeltz commented on service/geopmdpy/session.py:
```diff
@@ -31,91 +31,563 @@
+            geopm_proxy.PlatformOpenSession
+        except DBusError as ee:
+            if 'io.github.geopm was not provided' in str(ee):
+                err_msg = """The geopm systemd service is not enabled.
+    Install geopm service and run 'systemctl start geopm'"""
```
```suggestion
                err_msg = """The geopm systemd service is not enabled.
                             Install geopm service and run 'systemctl start geopm'"""
```

cmcantalupo replied:
We don't want that much white space in the printed message, I thought a 4 space indent looks okay.

bgeltz replied:
I really don't like the indentation.  Do we have this pattern any where else?

What about this instead?
```suggestion
                err_msg = "The geopm systemd service is not enabled. " \
                          "Install geopm service and run 'systemctl start geopm'"
```

bgeltz commented on service/geopmdpy/session.py:
```diff
@@ -31,91 +31,563 @@
+            RuntimeError: The period is non-zero or total time is negative.
+
+        """
+        if run_time <= 0.0:
```
I'm wondering if we ought to have some sort of private member, call it self._min_write_time or similar, that's the realistic minimum time for which a control can be applied.

This is going to be IO group control latency dependent, DBus latency dependent, etc.  I know it's greater than 0, but beyond that I'm not sure how to put a bound on it.

There should be docu that describes all these latencies and how they relate to this minimum run_time.

cmcantalupo replied:
I'm not sure what someone would be trying to achieve with a small, non-zero value, but it seems like user error if they expect that using a command line tool written in python is going to respond in microseconds.  

If the run_time is very small, we will set the controls, sleep for that period of time and then revert the values.  The latencies you are talking about are captured in the calls to geopm, not in the call to sleep.  We can add some documentation about how the run_time is the duration of time after all controls have been set until just before the process of reverting the values will begin.

bgeltz replied:
> We can add some documentation about how the run_time is the duration of time after all controls have been set until just before the process of reverting the values will begin.

This would clear it all up.  I'm just trying to avoid a situation where the time spent to apply the control is significantly greater than the sleep time, and the user being surprised they're not getting a period they ask for.

bgeltz commented on service/geopmdpy/session.py:
```diff
@@ -31,91 +31,563 @@
+
+        Iteration method for timed loop.  This iterator can be used in
+        a for statement to execute the loop periodically.
+
```
```suggestion
```

bgeltz commented on service/geopmdpy/session.py:
```diff
@@ -31,91 +31,563 @@
+        return [rr[0] for rr in self._requests]
+
+    def iterate_stream(self, request_stream):
+        """Iterated over a stream of requests
```
```suggestion
        """Iterate over a stream of requests
```

bgeltz commented on service/geopmdpy/session.py:
```diff
@@ -31,91 +31,563 @@
+                yield line
+
+
+class ReadRequests(Requests):
```
We're doing plurals now?

cmcantalupo replied:
The main purpose is to iterate over the object, so plural seemed appropriate.  We could call it a ReadRequestQueue?

bgeltz replied:
Fine w/either.  I just want to be consistent with all our code and I know we've been back and forth on having plural names in the codebase in the past.

bgeltz commented on service/geopmdpy/session.py:
```diff
@@ -31,91 +31,563 @@
+                      geopm_proxy.PlatformGetSignalInfo(signal_names)]
+        except DBusError as ee:
+            if 'io.github.geopm was not provided' in str(ee):
+                err_msg = """The geopm systemd service is not enabled.
+    Install geopm service and run 'systemctl start geopm'"""
```
```suggestion
                err_msg = """The geopm systemd service is not enabled.
                             Install geopm service and run 'systemctl start geopm'"""
```

cmcantalupo replied:
We don't want that much white space in the printed message, I thought a 4 space indent looks okay.

bgeltz replied:
What about this instead?
```suggestion
                err_msg = "The geopm systemd service is not enabled. " \
                          "Install geopm service and run 'systemctl start geopm'"
```

bgeltz reviewed (changes requested):
Looks pretty good overall.

cmcantalupo committed:
Brad's suggestions from code review for session.py

Co-authored-by: Brad Geltz <brad.geltz@intel.com>

cmcantalupo committed:
Change name of Requests objects to RequestQueue

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add sphinx documentation for the service command line tools.

- Add sections to reference rst file.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo commented on service/docs/source/reference.rst:
```diff
@@ -2,6 +2,64 @@
+    MSR::PERF_CTL:FREQ core 0 2.0e9
+
+
+The time specified is the duration after the write has completed until
```
@bgeltz Doc for latency is here.
